### ISSUE TITLE
Consider supporting `FOO=bar cmd` syntax

### ISSUE THREAD

**Issue by lilyball** [fish-shell/fish-shell#6048]
We should really reconsider supporting `FOO=bar cmd` syntax, as shorthand for `begin; set -lx FOO bar; cmd; end`. This is one of the more commonly-used sh-isms that Fish doesn't support and it's irritating every time I hit it.

Also see #438 wherein `env` doesn't support fish functions, so `env FOO=bar someFishFunc` doesn't work. There's some discussion in that ticket about turning `env` into a builtin just to solve this problem, but it's not clear why we don't want to just support `FOO=bar cmd`. This syntax does preclude having `=` be part of an executable name, but that ship already sailed ages ago with sh (not to mention it's rather weird), and that can be supported anyway with the `command` builtin if absolutely necessary (just like on `sh`).

**Comment by floam** [comment]
Let’s do it.

**Comment by filiden** [comment]
Is it not a break with certain principles of conception?

> There should only be one type of input to the shell, lists of commands. Loops, conditionals and variable assignments are all performed through regular commands.

> Fish should be user friendly, but not at the expense of expressiveness. Most tradeoffs between power and ease of use *can be avoided with careful design*.

**Comment by filiden** [comment]
Can you describe a general use case and summarize the issue? The links provided may be hard to follow because issues may have been improved (comments [about some issues](https://github.com/fish-shell/fish-shell/issues/438) are made since 2012).

POSIX does not seem to define this feature. Moreover, the environment defined by environment variables can be temporarily modified before the command execution using `env`.

**Comment by ridiculousfish** [comment]
I'm open to adding this. Some questions that should be answered before we can implement anything:

1. What is allowed to appear on the left and right side of the =?

2. What expansions are performed on both sides (tildes, variables, globs, command substitutions?) and what happens when the results are arrays?

3. When does the expansion happen? In particular `x=5 echo $x` does not work in bash. Is there a good reason for that decision, and what should fish do?

Of course the main motivation here is to be familiar to users coming from sh, but IIRC bash has especially counterintuitive semantics here when you dig into it so we should be a little careful.

**Comment by lilyball** [comment]
1. I would suggest we parse a single token, then check to see if it starts with a valid variable identifier followed by an `=`. Open questions: What if I quote this token, e.g. `"foo=23" cmd`? A quick test actually shows that in Bash this disables this syntax and tries to invoke `foo=23` as a command. I'm tempted to say we should follow suite, because `"FOO=23" cmd` doesn't obviously look like it should behave the same as `FOO=23 cmd`.

   Note that this token rule means the following actually is valid:

   ```
   FOO=one\
   two cmd
   ```

   Personally, I'm ok with that. Heck, it looks like this actually works in Bash too. In fact, so does putting the escaped newline prior to the `=`; I suspect Bash is stripping this escaped newline before doing any parsing. I don't know if we need to support that particular edge case though.

   I guess this rule can be more simply expressed as: If the first token's *raw string* (i.e. no tokenization) begins with a valid variable identifier followed by an equals sign, then treat it as a temporary variable assignment (and then run this check again on the next token, of course).

2. Since the LHS is "valid variable identifier" that means no expansion. We could argue for supporting `$foo=BAR` but I'd rather say "don't do that" for now. We can always revisit this later if there's a compelling reason to support it.

   For the RHS, we should run the normal set of expansions, which is to say `FOO=bar` should produce the exact same variable that `set -lx FOO bar` does, including command substitutions doing newline-splitting to produce a multi-valued variable. This behavior seems the least surprising (since it can be literally translated into `set -lx` without changing behavior).

3. This is a very good question. It bugs me that `x=5 echo $x` doesn't work in Bash. I'd like to break with precedent here and do the variable expansion first, so that way this behaves pretty much identically to `begin; set -lx x 5; echo $x; end`. This seems more useful (I've many times wanted `FOO=bar echo $FOO` to work, but I don't think I've ever been thankful for the current behavior).

Regardless of the actual implementation, I also think that the temporary variable's flavor should be a "local exported" variable, so e.g. `FOO=bar set -qlx FOO` should succeed.

**Comment by mqudsi** [comment]
> Let’s do it. There will practically never be a command with “=“ in it intended to be used by a shell, because of decades of other shells.

Yes, but we should be careful to not break literal arguments with `=` in their text. It's probably sufficient to do something like the following (terminating as soon as a non `env` argument is encountered).

```
for (auto i = 0; i < args.size() && args[i].find("=") != string::npos; ++i) {
    // Split and assign
}
```

But the devil is, as they say, always in the details. e.g. re @lilyball's suggestion

> Regardless of the actual implementation, I also think that the temporary variable's flavor should be a "local exported" variable, so e.g. FOO=bar set -qlx FOO should succeed.

I agree that command should succeed. **But** it would be a very grave mistake for it to enter the *existing* local scope.

```fish
FOO=bar set -qlx FOO && echo "set once"
set -qlx && echo "set twice"
```

should echo (for obvious reasons)

```
set once
```

It would be horrible if any time you wanted to set an environment variable for a single command it ended up being set for the current session/scope, e.g.

```fish
PATH=/foo bar
# everything goes wrong here if that was exported to the actual local scope
```

but that implies the creation of a *new* scope, which means this *wouldn't* do what you want:

```
FOO=bar set -qlx FOO && set -l bar "bar"
echo $bar
```

(back to our old problem with eval)

**Comment by lilyball** [comment]
> Yes, but we should be careful to not break literal arguments with = in their text.

Well no, we won't break those, because env vars must be specified prior to any other token (including the command itself).

> But it would be a very grave mistake for it to enter the existing local scope.

That would also completely violate the feature as described and would break with precedent. There's a reason the issue description establishes a new `begin … end` block.

**Comment by lilyball** [comment]
Another thing just occurred to me to specify: Env vars specified this way only apply to a single command, not to an entire pipeline. And they can be specified for commands in the middle of a pipeline. This matches Bash behavior. By that I mean

```fish
FOO=bar env | grep FOO # this works
true | FOO=bar env | grep FOO # this works too
FOO=bar true | env | grep FOO # this doesn't work
```

If you want it to apply to an entire pipeline you can use a `begin` block, as in

```fish
FOO=bar begin; true | env | echo FOO; end
```

**Comment by mqudsi** [comment]
Sorry, I didn't mean to imply that you hadn't taken it into account so much as it was a note for ourselves for future reference.

The issue with the invisible scope declaration remains, though. It is subtle enough to cause problems and confusion, as we previously experienced with `eval`.

**Comment by lilyball** [comment]
Here's a question: What's the behavior of `FOO=bar not true | env`? Since `not` negates the whole pipeline. Should we even allow this, or should we require `not FOO=bar true | env` instead?

**Comment by floam** [comment]
Just stating the obvious, we should want to be able to do `var1=a var2=b /bin/foo` as well.

zsh allows array indices:
```shell
$ zsh
$ export var=abc
$ var[2]=z env|grep var
var=azc
```

**Comment by lilyball** [comment]
If we were to support indices it should work exactly like it does with `set` (i.e. setting an entry of an array variable), except of course you can only include one index because you're only providing one value. But indices is something we can avoid supporting for now and add support later if we decide it's worth it.

**Comment by floam** [comment]
Well, it could be meaningful with colon-delimited vars like PATH.

**Comment by ridiculousfish** [comment]
@lilyball 's suggestion of interpreting only a "raw string" where the left side is a variable value sounds good to me. We can always enable expanding the left side later. I think I will also defer index expansion too initially.

I agree with breaking with bash and expanding variables before expanding the command.

The scope issue is well-spotted. It would be straightforward for `ENV=VAL cmd` to create a "hidden" local scope which is inaccessible to `set -l`.

**Comment by krobelus** [comment]
This should be quite convenient for many users.

> The scope issue is well-spotted. It would be straightforward for `ENV=VAL cmd` to create a "hidden" local scope which is inaccessible to `set -l`.

I have a [branch](https://github.com/krobelus/fish-shell/commit/variable-assignments) that implements this. It makes the variables accessible to `set -l` at the cost of being not that straightforward.
It seems to work as expected for all cases mentioned in this thread, differing from bash mainly in the fact that `FOO=bar echo $FOO` does print `bar` because it is implemented to work like `begin; set -lx FOO bar; echo $FOO; end`.

As an alternative we could abuse the abbreviation engine to expand `foo=bar` to `env foo=bar` which would probably be a simpler change. This could be also be done by expanding abbreviations based on user defined-regexes which sounds both tempting and scary.

**Comment by lilyball** [comment]
Abusing the abbreviation engine for this strikes me as a bad idea.

Besides, as I believe I said before, I actually like the idea that `FOO=bar echo $FOO` echoes "bar"; the fact that bash doesn't behave this way has only ever served as a source of irritation.

**Comment by zx8** [comment]
I'm 100% with @lilyball on this one. I always felt that bash missed a trick with `FOO=bar echo $FOO` _not_ outputting "bar", and would love if fish did not inherit bash's behaviour here and _did_ output "bar".
